### PR TITLE
fix: when bucket migrate conflict, migrate ec piece 's bug

### PR DESCRIPTION
**Pull request by jingjunLi**
### Description

fix some issues when bucket migrate conflict, migrate ec piece 's bug

### Rationale

Fixed Issues:
* Adapted the replicateHandler function to accommodate the bucket migration process, where the destination side no longer needs to verify that the sender is necessarily the primary service provider.

* Resolved conflicts during bucket migration:
  * Addressed the problem of not obtaining the latest spID.
  * Handled errors during destination GVG  transmission, preventing automatic skipping.
  * Set `redundancyIdx` and `finished` flags correctly in the `doneBucketMigrationReplicatePiece` function.

### Example

NA

### Changes

Notable changes: 
* Enhanced logging for better issue visibility.
* ...

### CODE REVIEW

File: modular/executor/migrate_task.go [hunk 1]
```diff
@@ -182,19 +191,19 @@ func (e *ExecuteModular) checkGVGConflict(ctx context.Context, srcGvg, destGvg *
...
 		err = e.doBucketMigrationReplicatePiece(ctx, destGvg.GetId(), objectInfo, params, spInfo.GetEndpoint(), segIdx, uint32(index), pieceData)
 		if err != nil {
-			log.CtxErrorw(ctx, "failed to do bucket migration to replicate pieces", "error", err)
+			log.CtxErrorw(ctx, "failed to do bucket migration to replicate pieces", "error", err, "piece_key", pieceKey)
 			return err
 		}
 	}
 
-	err = e.doneBucketMigrationReplicatePiece(ctx, destGvg.GetId(), objectInfo, params, spInfo.GetEndpoint(), uint32(index))
+	err = e.doneBucketMigrationReplicatePiece(ctx, destGvg.GetId(), objectInfo, params, spInfo.GetEndpoint(), 0, uint32(index))
```
Comment (will-2012) [inline modular/executor/migrate_task.go:204]: why here is 0.

File: modular/receiver/receive_task.go [hunk 2]
```diff
@@ -65,6 +65,7 @@ func (r *ReceiveModular) HandleReceivePieceTask(ctx context.Context, task task.R
 	defer r.receiveQueue.PopByKey(task.Key())
 	checksum := hash.GenerateChecksum(data)
 	if !bytes.Equal(checksum, task.GetPieceChecksum()) {
+		log.CtxErrorw(ctx, "failed to compare checksum", "task", task, "generate_checksum", checksum, "task_checksum", task.GetPieceChecksum())
```
Comment (will-2012) [inline modular/receiver/receive_task.go:68]: actual_checksum and expected_checksum is more meanful.

File: modular/executor/migrate_task.go [hunk 3]
```diff
@@ -219,21 +228,23 @@ func (e *ExecuteModular) doBucketMigrationReplicatePiece(ctx context.Context, gv
...
 		log.CtxErrorw(ctx, "failed to replicate piece", "segment_piece_index", segmentIdx,
 			"redundancy_index", redundancyIdx, "error", err)
+		return err
 	}
 	return nil
 }
 
 func (e *ExecuteModular) doneBucketMigrationReplicatePiece(ctx context.Context, gvgID uint32, objectInfo *storagetypes.ObjectInfo,
-	params *storagetypes.Params, destSPEndpoint string, segmentIdx uint32) error {
+	params *storagetypes.Params, destSPEndpoint string, segmentIdx, redundancyIdx uint32) error {
```
Comment (will-2012) [inline modular/executor/migrate_task.go:237]: Remove this `segmentIdx`

File: modular/executor/migrate_task.go [hunk 4]
```diff
@@ -219,21 +228,23 @@ func (e *ExecuteModular) doBucketMigrationReplicatePiece(ctx context.Context, gv
...
 	receive := &gfsptask.GfSpReceivePieceTask{}
-	receive.InitReceivePieceTask(gvgID, objectInfo, params, coretask.DefaultSmallerPriority, segmentIdx, -1, 0)
+	receive.InitReceivePieceTask(gvgID, objectInfo, params, coretask.DefaultSmallerPriority, segmentIdx, int32(redundancyIdx), 0)
```
Comment (will-2012) [inline modular/executor/migrate_task.go:239]: receive.InitReceivePieceTask(gvgID, objectInfo, params, coretask.DefaultSmallerPriority,`0/* this arg useless*/`, int32(redundancyIdx), 0)

**Review by will-2012 — APPROVED** [review]

File: modular/executor/migrate_task.go [hunk 5]
```diff
@@ -215,36 +224,38 @@ func (e *ExecuteModular) doBucketMigrationReplicatePiece(ctx context.Context, gv
 			"redundancy_index", redundancyIdx, "error", err)
 		return err
 	}
+	if int(segmentIdx+1) >= len(objectInfo.GetChecksums()) {
```
Comment (will-2012) [inline modular/executor/migrate_task.go:227]: redunct_index?

File: modular/executor/migrate_task.go [hunk 6]
```diff
@@ -215,36 +224,38 @@ func (e *ExecuteModular) doBucketMigrationReplicatePiece(ctx context.Context, gv
...
 		return err
 	}
 	receive.SetSignature(taskSignature)
 	receive.SetBucketMigration(true)
+	receive.SetFinished(true)
 	_, err = e.baseApp.GfSpClient().DoneReplicatePieceToSecondary(ctx, destSPEndpoint, receive)
 	if err != nil {
-		log.CtxErrorw(ctx, "failed to done replicate piece", "dest_sp_endpoint", destSPEndpoint,
-			"segment_idx", segmentIdx, "error", err)
+		log.CtxErrorw(ctx, "failed to done replicate piece", "dest_sp_endpoint", destSPEndpoint, "error", err)
```
Comment (will-2012) [inline modular/executor/migrate_task.go:254]: Ditto.

**Review by will-2012 — APPROVED** [review]

**Review by Fynnss — APPROVED** [review]

**Pull request merged**